LLama/Model/IModelCache.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using LLama.Common;
+
+namespace LLama.Model;
+
+/// <summary>
+/// A class that helps organize and load local models
+/// </summary>
+public interface IModelCache : IDisposable
+{
+    // Model Directories
+    /// <summary>
+    /// Configured set of directories that are scanned to find local models
+    /// </summary>
+    /// <value></value>
+    public IEnumerable<string> ModelDirectories { get; }
+
+    /// <summary>
+    /// Add a directory containing model files
+    /// </summary>
+    /// <param name="directory"></param>
+    public void AddDirectory(string directory);

```

Comment:
I have a few concerns about the abstractions here. As this API appears in `IModelCache`, it is required to be implemented in all its derived classes. However, they are likely to have similar implementations about this method, which may lead to duplicated code in the future. Some other APIs such as `RemoveDirectory`, `RemoveAllDirectories`, etc. are also in this case.

What about adding an abstract class that provides default implementations for these APIs? `ModelCache` -> `ModelCacheBase` -> `IModelCache`. The method `GetModelsFromDirectories`could also be placed in the base class.

On my side, a different implementations of this interface might be needed when we want to support remote file (for example, huggingface).

Reply:
That's a good point. To keep it as a minimal as possible the `IModelCache` interface should probably just have:
 - Get a model
 - Get list of all loaded models
 - Get available models (maybe)

Reply:
You both make good points, let me tweak this a bit

Reply:
I still need to finalize the shared semantics a bit. I reworked the interface(s) a bit. Let me know what you think of the separation